pkg/lib/indexer/indexer.go

```diff
@@ -573,16 +590,103 @@ type bundleDirPrefix struct {
 	pkgName, bundleVersion string
 }
 
-func getBundlesToExport(dbQuerier pregistry.Query, packages []string) (map[string]bundleDirPrefix, error) {
-	bundleMap := make(map[string]bundleDirPrefix)
+type bundleExportInfo struct {
+	name       string
+	bundlePath string
+	replaces   string
+	skips      map[string]struct{}
+}
 
+func getBundlesToExport(dbQuerier pregistry.Query, packages []string) (map[bundleDirPrefix]bundleExportInfo, error) {
+	bundleMap := make(map[bundleDirPrefix]bundleExportInfo)
+	graphLoader := &sqlite.SQLGraphLoader{
+		Querier: dbQuerier,
+	}
 	for _, packageName := range packages {
-		bundlesForPackage, err := dbQuerier.GetBundlesForPackage(context.TODO(), packageName)
+		bundleGraph, err := graphLoader.Generate(packageName)
 		if err != nil {
 			return nil, err
 		}
-		for k, _ := range bundlesForPackage {
-			bundleMap[k.BundlePath] = bundleDirPrefix{pkgName: packageName, bundleVersion: k.Version}
+		versionMap := make(map[string]string)
+		for _, channel := range bundleGraph.Channels {
+			var replaceDepth map[string]int64
+			for node, replaceNodes := range channel.Nodes {
+				bundleKey := bundleDirPrefix{
+					pkgName:       packageName,
+					bundleVersion: node.Version,
+				}
+				var replaces string
+				var skips map[string]struct{}
+				if bundleInfo, ok := bundleMap[bundleKey]; ok {
+					if bundleInfo.name != node.CsvName || bundleInfo.bundlePath != node.BundlePath {
+						return nil, fmt.Errorf("cannot export bundle version %s for package %s: multiple CSVs found", node.Version, packageName)
+					}
+					replaces = bundleInfo.replaces
+					skips = bundleInfo.skips
+				}
+				if skips == nil {
+					skips = map[string]struct{}{}
+				}
+				if len(replaces) > 0 {
+					skips[replaces] = struct{}{}
+				}
+				for replace := range replaceNodes {
+					if replace.CsvName == replaces {
+						continue
+					}
+
+					skips[replace.CsvName] = struct{}{}
+
+					if _, ok := channel.Nodes[replace]; !ok {
+						//synthetic entry
+						continue
+					}
+					if len(replaces) == 0 {
+						replaces = replace.CsvName
+						versionMap[replaces] = replace.Version
+						continue
+					}
+
+					var err error
+					if replaceDepth == nil {
+						replaceDepth, err = dbQuerier.GetBundleReplacesDepth(context.TODO(), packageName, node.CsvName)
+					}
+					if err == nil && replaceDepth[replaces] != replaceDepth[replace.CsvName] {
+						if replaceDepth[replaces] > replaceDepth[replace.CsvName] {
+							replaces = replace.CsvName
+							versionMap[replaces] = replace.Version
+						}
+						continue
+					}
+
+					// 2 replaces versions at the same depth on replaces chains of different channels
+					// Choose the newer semver as the replaces
+					replacesSemver, err := semver.Make(versionMap[replaces])
```

Comment:
As far as I know, there shouldn't be any ambiguity in the edges stored in the index that needs to be resolved by something like version comparison.